R/bootstrap-model.R

```diff
@@ -0,0 +1,668 @@
+){
+
+  checkmate::assert_class(.mod, NM_MOD_CLASS)
+
+  model_dir <- get_model_working_directory(.mod)
+
+  boot_run <- copy_model_from(
+    .parent_mod = .mod,
+    .new_model = .suffix,
+    .add_tags = "BOOTSTRAP_SUBMISSION",
+    .inherit_tags = .inherit_tags,
+    .update_model_file = TRUE,
+    .overwrite = .overwrite
+  )
+
+  boot_run[[YAML_MOD_TYPE]] <- "nmboot"
+  boot_run <- save_model_yaml(boot_run)
+
+  # Change problem statement
+  prob <- glue("Bootstrap run of model {get_model_id(.mod)}")
+  modify_prob_statement(boot_run, prob)
+
+  # Optionally remove $TABLE and $COV statements here
+  if(isTRUE(remove_cov)) remove_records(boot_run, type = "covariance")
+  if(isTRUE(remove_tables)) remove_records(boot_run, type = "table")
+
+  # Return read-in model to get the updated class as part of the model object
+  return(read_model(file.path(model_dir, .suffix)))
+}
+
+
+#' Set up a bootstrap model run.
+#'
+#' Creates a new model object and re-sampled dataset per model run.
+#'
+#' @param .boot_run a `bbi_nmboot_model` object.
+#' @param n number of model runs.
+#' @param strat_cols columns to maintain proportion for stratification
+#' @param seed a seed for sampling the data
+#' @param .overwrite logical (T/F) indicating whether or not to overwrite existing
+#'  setup for a bootstrap run.
+#'
+#' @seealso new_bootstrap_run summarize_bootstrap_run
+#'
+#' @examples
+#' \dontrun{
+#'
+#' # Setup
+#' .boot_run <- new_bootstrap_run(.mod)
+#' .boot_run <- setup_bootstrap_run(.boot_run, n = 1000, seed = 1234)
+#'
+#' # Submit
+#' submit_model(.boot_run)
+#'
+#' # Check status of runs during submission
+#' get_model_status(.boot_run)
+#' }
+#' @export
+setup_bootstrap_run <- function(
+    .boot_run,
+    n = 100,
+    strat_cols = NULL,
+    seed = 1234,
+    .overwrite = FALSE
+){
+  # TODO: make sure all _relevant_ resample_df args are part of this function
+  checkmate::assert_class(.boot_run, NMBOOT_MOD_CLASS)
+  checkmate::assert_number(n, lower = 1)
+  checkmate::assert_number(seed)
+
+  boot_dir <- get_output_dir(.boot_run, .check_exists = FALSE)
+  boot_data_dir <- file.path(boot_dir, "data")
+
+  # Bootstrap directory setup
+  if(!fs::dir_exists(boot_dir) || .overwrite == TRUE){
+    if(fs::dir_exists(boot_dir)) fs::dir_delete(boot_dir)
+    fs::dir_create(boot_dir)
+    fs::dir_create(boot_data_dir)
+    # Store data within run folder and gitignore
+    ##### TODO: UPDATE TO IGNORE MODEL FILES...
+    writeLines("/data\n*.ctl\n*.mod\n*.yaml", file.path(boot_dir, ".gitignore"))
+
+    # New model setup
+    n_seq <- seq(n)
+    mod_names <- purrr::map_chr(n_seq, max_char = nchar(n), pad_left)
+    mod_paths <- file.path(boot_dir, mod_names)
+    orig_mod <- read_model(get_based_on(.boot_run))
+
+    # TODO: only include subjects that entered the original problem
+    starting_data <- nm_data(orig_mod) %>% suppressMessages()
+    if(!is.null(strat_cols)){
+      checkmate::assert_true(strat_cols %in% names(starting_data))
```

Comment:
This currently fails if there are more than one stratification column 

```suggestion
      checkmate::assert_true(all(strat_cols %in% names(starting_data)))
```

Reply:
@barrettk making sure you don't miss this one